fix(node): Remove old committee members

gokhan-simsek-iota opened the pull request:
# Description of change

`fn handle_trusted_peer_change_event` was handling `TrustedPeerChangeEvent.new_peers` by always inserting the new trusted peers to known peers list, without removing the old ones.
With this PR, we update the known peers by removing the old peers and adding the latest trusted new peers and we do this without deleting the allowlisted peers which might have the latest PeerInfo (which might have come from `TrustedPeerChangeEvent.new_peers` with the same PeerId).

The changes require a new `batch_update` function in anemo, so the version is updated in Cargo.toml.
https://github.com/MystenLabs/anemo/pull/66 

## Links to any relevant issues

https://github.com/iotaledger/iota/issues/4902

## Type of change

- Enhancement (a non-breaking change which adds functionality)
 
## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

gokhan-simsek-iota committed:
First version

gokhan-simsek-iota committed:
Second solution, without using last_epoch_committee

gokhan-simsek-iota committed:
Typo

gokhan-simsek-iota committed:
Comments

gokhan-simsek-iota committed:
Format fix

gokhan-simsek-iota committed:
Comments

gokhan-simsek-iota committed:
Comment

muXxer commented on crates/iota-network/src/discovery/mod.rs:
```diff
@@ -212,18 +212,36 @@ impl DiscoveryEventLoop {
+        // the allowlisted peers.
+        // Note: self.network.known_peers consists of:
+        // self.allowlisted_peers + TrustedPeerChangeEvent.new_peers
+        let _: Vec<_> = self.network.known_peers().remove_all().collect();
```
Wouldn't that introduce race conditions, in case someone else accesses the known_peers `Arc<RwLock<HashMap<PeerId, PeerInfo>>>` in the meantime?

Wouldn't it be better to add an `update` function to the `anemo::KnownPeers` with peers_to_add, peers_to_remove?

gokhan-simsek-iota replied:
It would be if that was shared between tasks.

Firstly, `DiscoveryEventLoop.network.known_peers` is used in one task (`DiscoveryEventLoop` task) so the data race is not possible in current design/usage.
https://github.com/iotaledger/iota/blob/a96d831e3b8d0e07471f50a11ad851ae0618294d/crates/iota-network/src/discovery/builder.rs#L164

Secondly, the API of `anemo::KnownPeers` (or its parent `anemo::Network`) does not allow to acquire lock which suggests that it is not designed to be used concurrently from multiple execution contexts (in our case tokio tasks).

Even if we wanted to protect it from future concurrent usage (multiple tasks), we will get the this error:
```
    fn handle_trusted_peer_change_event()
    {
        // Acquire write lock once and hold it
        let mut known_peers = self.network.known_peers().0.write().unwrap();
            // error[E0616]: field `0` of struct `KnownPeers` is private
        // ... do things with `known_peers`
    } // Lock is automatically released when `known_peers` local var goes out of scope
```

The lock cannot be acquired, because inner of KnownPeers is private:
```
struct KnownPeers(Arc<RwLock<HashMap<PeerId, PeerInfo>>>);
```

If we really want to make it data race safe for possible future multi-task uses, we need API changes as you suggested, also need to take care of other places where `known_peers` used.

muXxer commented on crates/iota-network/src/discovery/mod.rs:
```diff
@@ -212,18 +212,36 @@ impl DiscoveryEventLoop {
+    /// the latest trusted new peers without deleting the allowlisted peers
     fn handle_trusted_peer_change_event(
         &mut self,
         trusted_peer_change_event: TrustedPeerChangeEvent,
```
Can't we figure out when the `TrustedPeerChangeEvent` gets created, which peers should be added and which should be removed? Ofc the "peers_to_remove" list should not remove peers that are part of the "allowlisted_peers". This is the initial list of trusted peers from the config, right?

semenov-vladyslav replied:
We figured it! The config contains allowlisted and seed peers; we implemented the event handling logic so that they can only be updated (if metadata changed) but not removed.

muXxer reviewed (changes requested):
I wonder if the meaning of `validators` changes in #4793.
Could it be that we only need to add the `committee` members to the trusted peers instead? @piotrm50 ? If that is the case, we need to change the logic.

Can we somehow utilize the emitted events from `validator_set.move`?

For example `ValidatorJoinEvent`, `ValidatorLeaveEvent`, `CommitteeValidatorJoinEvent`, `CommitteeValidatorLeaveEvent`.

jkrvivian committed:
feat: Add removed_peers to TrustedPeerChangeEvent

semenov-vladyslav committed:
fix: prettify

semenov-vladyslav committed:
feat: do not remove allowlisted known peers

jkrvivian committed:
feat: Move the validator diffs logic into DiscoveryEventLoop

gokhan-simsek-iota commented:
> I wonder if the meaning of `validators` changes in #4793. Could it be that we only need to add the `committee` members to the trusted peers instead? @piotrm50 ? If that is the case, we need to change the logic.
> 
> Can we somehow utilize the emitted events from `validator_set.move`?
> 
> For example `ValidatorJoinEvent`, `ValidatorLeaveEvent`, `CommitteeValidatorJoinEvent`, `CommitteeValidatorLeaveEvent`.

---

(1) About `committee_members`:

Looking at @piotrm50's PR more closely, in Move side, there will be `committee_members` **in addition** to `active_validators` in `ValidatorSetV2`:
https://github.com/iotaledger/iota/blob/e97851bd38e27cbbd7edf637b65eeb0b725795c8/crates/iota-framework/packages/iota-system/sources/validator_set.move#L70-L75

In current code (develop), `active_validators` are sent in every epoch start from `iota-node::send_trusted_peer_change()` to `iota-network :: DiscoveryEventLoop` by this fn in `EpochStartSystemStateV1`:

https://github.com/iotaledger/iota/blob/a96d831e3b8d0e07471f50a11ad851ae0618294d/crates/iota-types/src/iota_system_state/epoch_start_iota_system_state.rs#L227-L229

But, with @piotrm50's PR, `EpochStartSystemStateV1` will have `committee_members` **instead of** `active_validators`, so, in short, `iota-network :: DiscoveryEventLoop` will be receiving `committee_members` instead of `active_validators`, which seems to me does not effect this PR's logic if I understand your concern correctly assuming `committee_members` is what we need as trusted peers, not `active_validators`.

gokhan-simsek-iota commented:
> I wonder if the meaning of `validators` changes in #4793. Could it be that we only need to add the `committee` members to the trusted peers instead? @piotrm50 ? If that is the case, we need to change the logic.
> 
> Can we somehow utilize the emitted events from `validator_set.move`?
> 
> For example `ValidatorJoinEvent`, `ValidatorLeaveEvent`, `CommitteeValidatorJoinEvent`, `CommitteeValidatorLeaveEvent`.

(2) About those events:
Seems to me that those Move events are intended to be used by DApps front-ends, this looks very similar to Ethereum events (which I used before from a DApp).
https://docs.sui.io/guides/developer/sui-101/using-events#query-events-in-rust

jkrvivian committed:
Fix: Skip adding the node itself if it's in committee

jkrvivian committed:
fix: Insert new committee to known peers to update PeerInfo

semenov-vladyslav committed:
feat: simplify remove from known_peers

semenov-vladyslav committed:
feat: use send_modify to send TrustedPeerChangeEvent

jkrvivian committed:
fix: Fix return type of send_trusted_peer_change

jkrvivian committed:
test: Add unit test test_handle_trusted_peer_change_event

semenov-vladyslav committed:
fix: abbreviations for allowlisted and seed peers

semenov-vladyslav committed:
tests: clean up trusted peer change event test

semenov-vladyslav committed:
fmt

semenov-vladyslav committed:
cargo: bump anemo to the recent main commit

semenov-vladyslav committed:
feat: use anemo batch_update

semenov-vladyslav committed:
fmt

semenov-vladyslav committed:
test: added trusted peer change event test case

semenov-vladyslav commented:
Here is a brief summary of the updates:

- Added `old_committee` to `TrustedPeerChangeEvent`; the only long-term instance of this type is managed by `tokio::sync::watch` channel; this way we keep the new_committee from the previous epoch that becomes the old_committee in the new epoch; we use `send_modify` now to emit the event;
- the discovery loop removes the peers from old_committee that are not in new_committee and that are not allowlisted; then we add all new_committee peers (and thus probably update the allowlisted ones);
- switched to the latest anemo commit that introduces `batch_update` function needed for the atomic updates; that requires to update Cargo.lock as well;
- added `TrustedPeerChangeEvent` handling tests.

muXxer commented on Cargo.lock:
```diff
@@ -107,7 +107,7 @@ dependencies = [
  "once_cell",
  "version_check",
- "zerocopy",
+ "zerocopy 0.7.35",
```
Did you run `cargo update` on this branch?
IMO all those changes to this file need to be reverted, except the ones that update anemo.

semenov-vladyslav committed:
cargo: revert Cargo.lock and update anemo package only

gokhan-simsek-iota commented:
@semenov-vladyslav 
If we want to go down that road to make it data race safe, there is another multi step modify here:
https://github.com/iotaledger/iota/blob/df70cff8a2e68163d92d2057c8849dff58486d84/crates/iota-network/src/discovery/mod.rs#L206

So, we might need something like `insert_all()` in anemo, or maybe `batch_update` can be re-used.

jkrvivian commented:
> @semenov-vladyslav If we want to go down that road to make it data race safe, there is another multi step modify here:
> 
> https://github.com/iotaledger/iota/blob/df70cff8a2e68163d92d2057c8849dff58486d84/crates/iota-network/src/discovery/mod.rs#L206
> 
> So, we might need something like `insert_all()` in anemo, or maybe `batch_update` can be re-used.

This is called at the initialization phase (`configure_preferred_peers`), the discovery loop is not yet up, it should be safe as it is.

semenov-vladyslav committed:
cargo: fix update anemo-cli

jkrvivian committed:
chores: Fix an clippy error

jkrvivian committed:
chores: Fix ci-clippy errors

DaughterOfMars commented on crates/iota-network/src/discovery/mod.rs:
```diff
@@ -74,7 +74,8 @@ pub struct NodeInfo {
 /// Contains a new list of available trusted peers.
 pub struct TrustedPeerChangeEvent {
-    pub new_peers: Vec<PeerInfo>,
+    pub new_committee: Vec<PeerInfo>,
+    pub old_committee: Vec<PeerInfo>,
```
Any reason not to use HashSets here?

semenov-vladyslav replied:
We don't need it here, the type returned by `get_validator_as_p2p_peers` is Vec, PeerInfo is neither Eq, nor Hash.

gokhan-simsek-iota commented:
> > @semenov-vladyslav If we want to go down that road to make it data race safe, there is another multi step modify here:
> > https://github.com/iotaledger/iota/blob/df70cff8a2e68163d92d2057c8849dff58486d84/crates/iota-network/src/discovery/mod.rs#L206
> > 
> > So, we might need something like `insert_all()` in anemo, or maybe `batch_update` can be re-used.
> 
> This is called at the initialization phase (`configure_preferred_peers`), the discovery loop is not yet up, it should be safe as it is.

Yes, it is safe now, but the idea was to make data race proof for the future, `fn handle_trusted_peer_change_event` was safe (known_peers was not shared between tokio tasks) as it was but we wanted to make it data race safe for future possible uses. So what I mean is that  someone else can easily use `fn configure_preferred_peers` out of init phase in the future.

muXxer commented:
> Yes, it is safe now, but the idea was to make data race proof for the future.

Let's not over-engineer it. My point was that a batch_update function would make sense, because for me it was not super obvious if there could be data races or not on the first quick view of the code.
I didn't mean that every single part of the code needs to be future proof data race safe, even functions that are just used once during configuration phase. Let's be more pragmatic.

muXxer committed:
Merge branch 'consensus/feat/committee-selection-feature' of github.com:iotaledger/iota into node/remove-old-committee-members

muXxer committed:
fix: format

jkrvivian committed:
enhance(network): Add debug print to handle_trusted_peer_change_event

jkrvivian committed:
fix(network): Fix debug print in handle_trusted_peer_change_event

semenov-vladyslav committed:
fix: updated tracing output of changed known peers

muXxer commented:
The tests fail because of an incompatibility of the updated crates. We need to update those in a clean way in another PR first.

semenov-vladyslav committed:
cargo: update anemo and a few iota-tls deps

Inconsistent rustls-webpki and x509-parser versions in iota-tls and anemo crates cause simtest
test_full_node_sync_flood_determinism to fail:

https://github.com/iotaledger/iota/actions/runs/14034613903/job/39372674568?pr=5755
